### ISSUE TITLE
Neovim crashes when embedded and ui is attached with ext_cmdline enabled

### ISSUE THREAD

**Issue by vhakulinen** [neovim/neovim#8007]
- `nvim --version`: 
```
NVIM v0.2.2
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/home/ville/src/neovim/build/config -I/home/ville/src/neovim/src -I/home/ville/src/neovim/.deps/usr/include -I/home/ville/src/neovim/.deps/usr/include -I/home/ville/src/neovim/.deps/usr/include -I/home/ville/src/neovim/.deps/usr/include -I/home/ville/src/neovim/.deps/usr/include -I/home/ville/src/neovim/.deps/usr/include -I/usr/include -I/home/ville/src/neovim/build/src/nvim/auto -I/home/ville/src/neovim/build/include
Compiled by ville@Desktop

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"
```
- Operating system/version: Ubuntu 17.10

### Steps to reproduce using `nvim -u NORC`

Run some neovim gui with `ext_cmdline` enabled and type something along the lines of the following:
```viml
:function ASD()
endfunction
:function ASD()
endfunction
:endfunction
```
Neovim will crash once fuzzing enough with above lines (e.g. declare a function that is already declared, which results in neovim echoing some errors, and then trying to `:endfunction`).

Here is a diff to enable `ext_cmdline` for python-gui:

```diff
diff --git a/neovim_gui/ui_bridge.py b/neovim_gui/ui_bridge.py
index ae7d8af..954bb67 100644
--- a/neovim_gui/ui_bridge.py
+++ b/neovim_gui/ui_bridge.py
@@ -47,6 +47,7 @@ class UIBridge(object):
 
     def attach(self, columns, rows, **options):
         """Attach the UI to nvim."""
+        options["ext_cmdline"] = True
         self._call(self._nvim.api.ui_attach, columns, rows, options)
 
     def detach(self):
```

Here is a backtrace from a debug build from latest master:

```
[New LWP 19731]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/home/ville/src/neovim/build/bin/nvim --embed'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  je_arena_run_regind (bin_info=<optimized out>, bin_info=<optimized out>, ptr=<optimized out>, run=<optimized out>) at include/jemalloc/internal/arena.h:1167
1167		    bin_info->reg0_offset);
#0  je_arena_run_regind (bin_info=<optimized out>, bin_info=<optimized out>, ptr=<optimized out>, run=<optimized out>) at include/jemalloc/internal/arena.h:1167
#1  arena_run_reg_dalloc (ptr=<optimized out>, run=0x7fe619209de8) at src/arena.c:309
#2  arena_dalloc_bin_locked_impl (bitselm=<optimized out>, junked=true, ptr=<optimized out>, chunk=0x7fe619200000, arena=0x7fe619e00140, tsdn=0x7fe61b4b0730) at src/arena.c:2963
#3  je_arena_dalloc_bin_junked_locked (tsdn=tsdn@entry=0x7fe61b4b0730, arena=arena@entry=0x7fe619e00140, chunk=0x7fe619200000, ptr=<optimized out>, bitselm=<optimized out>) at src/arena.c:2981
#4  0x000055688e0979e6 in je_tcache_bin_flush_small (tsd=0x7fe61b4b0730, tcache=tcache@entry=0x7fe619c0d000, tbin=tbin@entry=0x7fe619c0d1a8, binind=<optimized out>, rem=16) at src/tcache.c:137
#5  0x000055688e053443 in je_tcache_dalloc_small (slow_path=false, binind=<optimized out>, ptr=0x7fe61928e600, tcache=<optimized out>, tsd=<optimized out>) at include/jemalloc/internal/tcache.h:419
#6  je_arena_dalloc (slow_path=false, tcache=<optimized out>, ptr=0x7fe61928e600, tsdn=<optimized out>) at include/jemalloc/internal/arena.h:1441
#7  je_idalloctm (slow_path=false, is_metadata=false, tcache=<optimized out>, ptr=0x7fe61928e600, tsdn=<optimized out>) at include/jemalloc/internal/jemalloc_internal.h:1170
#8  je_iqalloc (slow_path=false, tcache=<optimized out>, ptr=0x7fe61928e600, tsd=<optimized out>) at include/jemalloc/internal/jemalloc_internal.h:1187
#9  ifree (tsd=<optimized out>, slow_path=false, tcache=<optimized out>, ptr=0x7fe61928e600) at src/jemalloc.c:1896
#10 free (ptr=0x7fe61928e600) at src/jemalloc.c:2016
#11 0x000055688de9fe99 in xfree (ptr=0x7fe61928e600) at ../src/nvim/memory.c:133
#12 0x000055688dd734f8 in api_free_array (value=...) at ../src/nvim/api/private/helpers.c:889
#13 0x000055688dd73481 in api_free_object (value=...) at ../src/nvim/api/private/helpers.c:871
#14 0x000055688dd734d9 in api_free_array (value=...) at ../src/nvim/api/private/helpers.c:886
#15 0x000055688dd73481 in api_free_object (value=...) at ../src/nvim/api/private/helpers.c:871
#16 0x000055688dd734d9 in api_free_array (value=...) at ../src/nvim/api/private/helpers.c:886
#17 0x000055688deb9b6a in serialize_request (channel_id=1, request_id=0, method=..., args=..., sbuffer=0x55688e3847d0 <out_buffer>, refcount=1) at ../src/nvim/msgpack_rpc/channel.c:622
#18 0x000055688deb9311 in send_event (channel=0x7fe619cb3000, name=0x55688e0a64ed "redraw", args=...) at ../src/nvim/msgpack_rpc/channel.c:460
#19 0x000055688deb8042 in rpc_send_event (id=1, name=0x55688e0a64ed "redraw", args=...) at ../src/nvim/msgpack_rpc/channel.c:103
#20 0x000055688dd78235 in remote_ui_flush (ui=0x7fe619dc9a00) at ../src/nvim/api/ui.c:257
#21 0x000055688dfb332f in ui_call_flush () at src/nvim/auto/ui_events_call.generated.h:88
#22 0x000055688dfb5c96 in ui_flush () at ../src/nvim/ui.c:488
#23 0x000055688dea90b5 in msg_end () at ../src/nvim/message.c:2522
#24 0x000055688dea497e in msg_attr_keep (s=0x55688e383f60 <errbuf> "E122: Function OK3 already exists, add ! to replace it", attr=127, keep=0) at ../src/nvim/message.c:194
#25 0x000055688dea483a in msg_attr (s=0x55688e383f60 <errbuf> "E122: Function OK3 already exists, add ! to replace it", attr=127) at ../src/nvim/message.c:137
#26 0x000055688dea55ca in emsg (s_=0x55688e383f60 <errbuf> "E122: Function OK3 already exists, add ! to replace it") at ../src/nvim/message.c:575
#27 0x000055688dea5745 in emsgfv (fmt=0x55688e0a8da0 "E122: Function %s already exists, add ! to replace it", ap=0x7ffd2d6d6c30) at ../src/nvim/message.c:606
#28 0x000055688dea56ce in emsgf (fmt=0x55688e0a8da0 "E122: Function %s already exists, add ! to replace it") at ../src/nvim/message.c:590
#29 0x000055688ddbb590 in emsg_funcname (ermsg=0x55688e0a8da0 "E122: Function %s already exists, add ! to replace it", name=0x7fe6192c9ed8 "OK3") at ../src/nvim/eval.c:6444
#30 0x000055688dddb1bb in ex_function (eap=0x7ffd2d6d7030) at ../src/nvim/eval.c:19777
#31 0x000055688de1b2ec in do_one_cmd (cmdlinep=0x7ffd2d6d7230, flags=0, cstack=0x7ffd2d6d7340, fgetline=0x55688de34e05 <getexline>, cookie=0x0) at ../src/nvim/ex_docmd.c:2242
#32 0x000055688de179c5 in do_cmdline (cmdline=0x0, fgetline=0x55688de34e05 <getexline>, cookie=0x0, flags=0) at ../src/nvim/ex_docmd.c:609
#33 0x000055688dec6f72 in nv_colon (cap=0x7ffd2d6d79c0) at ../src/nvim/normal.c:4558
#34 0x000055688debf40d in normal_execute (state=0x7ffd2d6d7920, key=58) at ../src/nvim/normal.c:1136
#35 0x000055688df8aa6e in state_enter (s=0x7ffd2d6d7920) at ../src/nvim/state.c:67
#36 0x000055688debd6b2 in normal_enter (cmdwin=false, noexmode=false) at ../src/nvim/normal.c:466
#37 0x000055688de828e8 in main (argc=2, argv=0x7ffd2d6d7c78) at ../src/nvim/main.c:572
```

### Actual behaviour

Neovim crashes. This doesn't happen when in TUI or when `ext_cmdline` is not enabled with remove UI.

### Expected behaviour

Neovim shouldn't crash.

**Comment by bfredl** [comment]
Can reproduce with `-u NONE` (even if not with my default config). Seems double freed array, I will look into the relevant code.

(Also no need to intercept the option in the brigde, instead can just change the attach call in gtk_ui.py:98)

**Comment by vhakulinen** [comment]
> Seems double freed array

That might explain why I got invalid `redraw` messages randomly.

**Comment by bfredl** [comment]
@vhakulinen can you check if #8008 fixes the issue?

**Comment by vhakulinen** [comment]
Yup, that seems to fix it! Can't crash it anymore (with that branch) with the same fuzzing tests that I can use to crash the master branch and 0.2.2. version.

**Comment by bfredl** [comment]
Fixed by #8008
